pkg/lib/cockpit/dialog.tsx

```diff
@@ -0,0 +1,1543 @@
+
+    /* Unlike with the timeout, we can not cancel the old promise when
+       installing a new one. Instead we check at the end whether it is
+       still really us that is supposed to deliver the result, by
+       comparing promises.
+
+       To summarize:
+
+       - The round id check will fail if the value is no longer
+         relevant to the dialog.  For example, say there is a text
+         input that can be toggled in and out of the dialog via a
+         checkbox. Now a validation round is started while the text
+         input is part of the dialog. During the debounce timeout or
+         while the asynchronous validation function runs, the user
+         toggles the checkbox (which triggers a new validation round)
+         and the text input is no longer part of the dialog. Now when
+         the timeout or validation for the text input concludes, the
+         round id check fails and the result is ignored, as it should.
+
+       - The promise check will fail when a asynchronous validation
+         takes longer than the debounce timeout.  Let's say there is a
+         text input with a debounce timeout of 1 second and a
+         validation function that takes 2 seconds. The user makes a
+         change that triggers validation and then remains idle for
+         more than a second. After one second, the timeout expires and
+         the promise is created and starts running. It will finish at
+         second 3, but we are not there yet. At second 1.5 the user
+         makes another change, a new timeout expires at 2.5 and a new
+         promise is created. At second 3 the original promise finally
+         comes to a conclusion, and the path is still relevant to the
+         dialog, but this promise is no longer the current
+         promise. Its result will be ignored, as it should.
+     */
+
+    #validation_state_is_current(state: DialogValidationState, prom?: Promise<void>): boolean {
+        return (
+            (!prom || Object.is(state.promise, prom)) &&
+                this.#is_current_validation_round_id(state.round_id)
+        );
+    }
+
+    /* _validate_value_async puts this all together.
+     */
+
+    _validate_value_async(path: string, val: unknown, debounce: number, func: () => Promise<string | undefined>): void {
+        const state = this.#get_validation_state(path);
+        if (!this.#probe_validation_state_cache(state, val)) {
+            debug("async validate start debounce", state.path, val);
+            this.#set_validation_state_timeout(
+                state,
+                val,
+                debounce,
+                () => {
+                    debug("async validate start promise", state.path, val);
+                    const prom =
+                        func()
+                                .catch(
+                                    ex => {
+                                        console.error(ex);
+                                        return undefined;
```

Comment:
These 3 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-22700-2dd5f8c0-20251219-103022-fedora-43-devel-other/Coverage/lcov/github-pr.diff.gcov.html)